odoo/tools/misc.py

```diff
@@ -1629,6 +1633,69 @@ def hmac(env, scope, message, hash_function=hashlib.sha256):
     ).hexdigest()
 
 
+def hash_sign(env, scope, message_values, expiration=None, expiration_hours=None, hash_function=hashlib.sha256):
+    """ Generate an urlsafe payload signed with the HMAC signature for an iterable set of data.
+
+    :param env: sudo environment to use for retrieving config parameter
+    :param scope: scope of the authentication, to have different signature for the same
+        message in different usage
+    :param message_values: values to be encoded inside the payload
+    :param expiration: optional, a datetime or timedelta
+    :param expiration_hours: optional, a int representing a number of hours before expiration. Cannot be set at the same time as expiration
+    :return: the payload that can be used as a token
+    """
+    assert not (expiration and expiration_hours)
+    assert message_values is not None
+
+    if expiration_hours:
+        expiration = datetime.datetime.now() + datetime.timedelta(hours=expiration_hours)
+    else:
+        if isinstance(expiration, datetime.timedelta):
+            expiration = datetime.now() + expiration
+    expiration_timestamp = 0 if not expiration else int(expiration.timestamp())
+
+    # we need to encode in base64 individually before adding the separator to avoid any collision between the content and the separator.
+    message_strings = json.dumps(message_values)
+    message_strings_encoded = message_strings.encode()
+    b64_message_list = base64.urlsafe_b64encode(message_strings_encoded)
+    b64_message = b64_message_list.decode().rstrip('=')
+
+    hash_value = hmac(env, scope, f'{b64_message}:{expiration_timestamp}')
+
+    b64_expiration = base64.urlsafe_b64encode(expiration_timestamp.to_bytes(math.ceil(expiration_timestamp.bit_length() / 8), 'little')).decode().rstrip("=")
+    b64_hash = base64.urlsafe_b64encode(bytes.fromhex(hash_value)).decode().rstrip("=")
+
+    return f'1~{b64_message}~{b64_expiration}~{b64_hash}'
```

Comment:
shouldn't we include the version in the signature ? (otherwise in the future, if we have many version, we can change the version of a hash, and it might still be valid, depending on how it's checked)

Reply:
yes, but if we extend it, I guess we will have 2 different logic, based on the version
so from a specific token, you can use the other logic (so you sign a token with the version 1, and try to decode it with the version 2)
it really depends on how we extend it, so it's hard to say now if it will be unsafe or not, but I really think the version should be signed too (and it will break only if we try to modify the version)

Reply:
(so it's not in the message, but it still signed)

Reply:
good idea :) But I think we should put the version first, in case we change the hash function (and so its length) in a future version

Reply:
> oh but ofc it's not fixed length because we use unpadded base64 🤦

We can still do something like that
```py
b64(b"\1" + bytes.fromhex("hex hmac") + b"4 bytes timestamp" + json.dumps(data).encode())
```

(so we b64 encode the entire payload, so each part, except the message has a fixed length)

> So now I'm wondering: aren't we unnecessarily reinventing [JWT](https://jwt.io/)?

A bit yes, but the JWT take more place because of the header, so I guess it depend if the token size is important for us or not